### ISSUE TITLE
compilation error

### ISSUE THREAD

**Issue by cnpetra** [LLNL/hiop#47]
with `cmake -DHIOP_USE_MPI=NO ..` I get the compilation issue below:

```
[ 86%] Building CXX object tests/CMakeFiles/testMatrix.dir/testMatrix.cpp.o
In file included from /ccs/home/cpetra/work/projects/hiop/tests/LinAlg/matrixTestsDense.hpp:59:0,
                 from /ccs/home/cpetra/work/projects/hiop/tests/testMatrix.cpp:61:
/ccs/home/cpetra/work/projects/hiop/tests/LinAlg/matrixTests.hpp:63:20: fatal error: optional: No such file or directory
 #include <optional>
                    ^
compilation terminated.
make[2]: *** [tests/CMakeFiles/testMatrix.dir/testMatrix.cpp.o] Error 1
make[1]: *** [tests/CMakeFiles/testMatrix.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by cnpetra** [comment]
Wild guess: `<optional>` is only supported starting C++17, which in turn is not supported by the g++ from GNU 4.8.5 (not sure this is what cmake picks up -> I really did a `module load gcc/9.1.0`

**Comment by ashermancinelli** [comment]
I want to say we caught that earlier, but must not have removed the header from every file we used it in. I should do a better job of checking compatibility with the lower-bound of what we claim to support. I will also have a fix for this out soon.

**Comment by cnpetra** [comment]
imo, 4.8.5 is an unreasonably lower bound. :)

a quick fix
```
CC=/sw/summit/gcc/9.1.0-alpha+20190716/bin/gcc CXX=/sw/summit/gcc/9.1.0-alpha+20190716/bin/g++ cmake ..
```

**Comment by ashermancinelli** [comment]
I think we need some generic lambda features from gcc/4.9 and up. I think 5.0 is the lowest we should support (though I do wish we could count on all users using gcc/9.1.0 😄). I think they are technically c++14 features, but I don't think we have a choice if we're going to move forward with RAJA. Thoughts?

**Comment by pelesh** [comment]
@cnpetra when you load gcc module check what `which cc` and `which gcc` give you. Depending on configuration, CMake could pick up default system compiler and not look for `gcc`. In that case it suffices to configure with `CXX=g++ cmake ..`.

RAJA [needs](https://raja.readthedocs.io/en/master/getting_started.html) gcc >= 4.9 and CMake >= 3.9. I think it is reasonable to set gcc version requirement for HiOp to be >= 5.0.

**Comment by cnpetra** [comment]
I'm personally for, but I have to check... hiop is still used by some folks with gcc4.8.5.

**Comment by cnpetra** [comment]
gcc ver >= 5.0 fine

**Comment by ashermancinelli** [comment]
I believe this is resolved with #57

**Comment by cnpetra** [comment]
confirmed
